README.md
# IP2Location Dart Package
Verifies the source of ip connections and domains from IP2Location ip2location.io.

This Dart package is designed to enable you to verify geolocation and WHOIS information for IP addresses and domains. It is built on the IP2Location API, making it a powerful tool for understanding the location of personal connections.

[![Static Badge](https://img.shields.io/badge/pub-v1.0.3-blue)](https://pub.dev/packages/verifyip2location)

A Flutter ans dart plugin for get location and whois information.

|             | Android | iOS   | Linux | macOS  | Web | Windows     |
|-------------|---------|-------|-------|--------|-----|-------------|
| **Support** | SDK 16+ | 11.0+ | Any   | 10.14+ | Any | Windows 10+ |

## Features

    Retrieve geolocation data, including city name, region name, country name, and country code.
    Detect the use of proxies, Tor, or VPN.
    Get latitude and longitude coordinates of the location.
    Identify devices infected with botnets or malware.
    Fetch WHOIS information for domains.

## Usage


- Retrieving Geolocation Data

```
import 'package:verifyip2location/verifyip2location.dart';

void main() async {
  String ip2token = 'YOUR_IP2LOCATION_API_KEY';
  BaseHttpIkaResponse result = await getIpgeo(ip2token, ip: '8.8.8.8');
  
  // Use the resulting data as needed
  print('Country: ${result.countryName}');
  print('City: ${result.cityName}');
  // ...
}

```

 - Retrieving WHOIS Information for a Domain

 ```
 import 'package:verifyip2location/verifyip2location.dart';

void main() async {
  String ip2token = 'YOUR_IP2LOCATION_API_KEY';
  HttpIkaWhoisService whoisResult = await getDnsWhois(ip2token, domain: 'example.com');
  
  // Use the resulting WHOIS data as needed
  print('Domain Name: ${whoisResult.domain}');
  print('Domain Owner: ${whoisResult.registrantName}');
  // ...
}
```




To use this plugin, add `verifyip2location` as a [github repository](https://github.com/Hegeldirkk/verifyip2location)

### Example

Location Functionnality Description

<?code-excerpt "lib/verifyip2location.dart (basic-example)"?>
```dart
import 'package:verifyip2location/verifyip2location.dart' as verifyip2location;
import 'package:verifyip2location/verifyip2location.dart';



void main() async {
  // Your API key
  final apiKey = 'VOTRE_API_KEY';

  // ip2location class
  BaseHttpIkaResponse? ip2loc;

   // IP address you want to look up
  final ipAddress = '8.8.8.8';

  // Plan you want to use (e.g, 'free = 0', 'starter = 1', 'plus = 2', 'security = 3', etc.)
  int plan = 0;

  try {
    // Using the getIpgeo function to retrieve data
    ip2loc = await verifyip2location.getIpgeo(apiKey, ip: ipAddress, plan: plan, option: 'geoip');

    // Get information
    if (ip2loc.errorStatusCode == 10000 || ip2loc.errorStatusCode == 10001 ){
      print('Adresse IP: ${ip2loc.resIp}');
      print('Pays: ${ip2loc.countryName}');
      print('Ville: ${ip2loc.cityName}');
      print('Latitude: ${ip2loc.latitude}');
      print('Longitude: ${ip2loc.longitude}');
      print('As: ${ip2loc.as}');
      // And so on for other data you want to use.
    } else if (ip2loc.errorStatusCode == 10001){
      print('Error message: ' + ip2loc.errorReasonPhrase!);
    } else{
        print('Error message: ' + ip2loc.errorReasonPhrase!);
    }
  } catch (e) {
    print('An error occurred during the request : $e');
  }
}

```

Requirements

    Dart SDK 2.0 or later.

Notes

    To use this package, you need to obtain an IP2Location API key from the official IP2Location website.
    Ensure you configure your API key in your application before using this package.

Author

This package was created by [DIABAKATE IKARY RYANN](https://dirkk.tech).

CHANGELOG.md
## [1.0.0-dev.1] - 2023-09-29

- Initial version.


test/verifyip2location_test.dart
import 'package:verifyip2location/src/base_response_whois.dart';
import 'package:verifyip2location/verifyip2location.dart' as verifyip2location;
import 'package:verifyip2location/src/base_response.dart';
import 'package:test/test.dart';

void main() {
  group('A group of tests', () {
    String? codecounrty;
    BaseHttpIkaResponse? ip2loc;

    setUp(() async {
      // Effectuer des opérations asynchrones ici si nécessaire.

      ip2loc = await verifyip2location.getIpgeo(
          "EF3080DDEA1CBF08D4BA41DEE162DE54",
          plan: 0,
          option: "geoip",
          ip: "191.252.201.137.");

      codecounrty = ip2loc!.countryCode;
    });

    test('Test if response is a Notnull', () {

      if (ip2loc!.errorStatusCode == 10000 || ip2loc!.errorStatusCode == 10001 ){
        print(ip2loc!.errorReasonPhrase);
      }

      expect(codecounrty, isNotNull);
      print(ip2loc!.as);
      expect(ip2loc!.elevation, isNull);
      print(ip2loc!.elevation);
    });
  });

  group('test whois', () {
    String? codecounrty;
    HttpIkaWhoisService? ip2loc;

    setUp(() async {
      // Effectuer des opérations asynchrones ici si nécessaire.

      ip2loc = await verifyip2location.getDnsWhois(
          "EF3080DDEA1CBF08D4BA41DEE162DE54",
          plan: 0,
          option: "whois",
          domain: "dirkk.tech");
      codecounrty = ip2loc!.registrant.country;
    });

    test('Test if response is a Notnull', () {
      //print(ip2loc);
      if (ip2loc!.errorCode == 10001) {
        print(ip2loc!.errorMessage);
      } else {
        expect(codecounrty, isNotNull);
        print(ip2loc!.nameServer);
      }
    });
  });
}


lib/src/ip2config.dart
///
///Geolocation API Endpoint Configuration
///@plan: var plan of IP2Location.io
///@url: endpoint url
///@apikey: apikey of IP2Location to use services

enum Plan { free, starter, plus, security }

class ConfigIP {
  String? plan;
  String? apikey;
  final String url = "https://api.ip2location.io/?";
  final String urlW = "https://api.ip2whois.com/v2?";

  String? choosePlan(int chp) {

    print(chp);
    // verify that chp is not negative
    assert(chp >= 0, "chp variable must be positive ");

    // check that chp is less than 4
    assert(chp <= 4, "chp variable must be less than 4 ");


    // assign the plan
    switch (chp) {
      case 0:
        plan = Plan.free.name;
      case 1:
        plan = Plan.starter.name;
      case 2:
        plan = Plan.plus.name;
      case 3: 
       plan = Plan.security.name;
        break;
      default:
        plan = Plan.free.name;
    }
    
    return plan;
  }
}


///
///


lib/src/base_response.dart
/// base class for http response
///
///

class BaseHttpIkaResponse {
  /// IP2Location error status code
  late int errorStatusCode;

  /// Raison Phrase error status code
  late String? errorReasonPhrase;

  /// IP Address response data
  /// IPv4 : 8.8.8.8
  late String? resIp;

  /// Country code response data
  /// Two-character country code based on ISO 3166.
  late String? countryCode;

  /// Country name response data
  /// Country name based on ISO 3166.
  late String? countryName;

  /// Region name response data
  /// Region or state name.
  late String? regionName;

  /// City name response data
  late String? cityName;

  /// Latitude and longitude Location information
  /// Geo coordinates
  /// latitude data location
  /// City latitude. Defaults to capital city latitude if city is unknown.
  late double? latitude;

  /// longitude data location
  /// City longitude. Defaults to capital city longitude if city is unknown.
  late double? longitude;

  /// Zip code response data
  /// ZIP/Postal code.
  late String? zipCode;

  /// Time zone response data
  /// UTC time zone (with DST supported).
  late String? timeZone;

  /// Asn response data
  /// Autonomous system number (ASN).
  late String? asn;

  /// As response data
  /// Autonomous system (AS) name.
  late String? as;

  /// Proxy response data
  /// Whether is a proxy or not
  late bool isProxy;

  // Nouvelles propriétés ajoutées pour les données complexes
  late String? isp;
  late String? domain;
  late String? netSpeed;
  late String? iddCode;
  late String? areaCode;
  late String? weatherStationCode;
  late String? weatherStationName;
  late String? mcc;
  late String? mnc;
  late String? mobileBrand;
  late int? elevation;
  late String? usageType;
  late String? addressType;
  late Continent continent; // Classe Continent
  late Country country; // Classe Country
  late Region region; // Classe Region
  late City city; // Classe City
  late TimeZoneInfo timeZoneInfo; // Classe TimeZoneInfo
  late Geotargeting geotargeting; // Classe Geotargeting
  late Proxy proxy; // Classe Proxy
  late String? adsCategory;
  late String? district;

  BaseHttpIkaResponse(
      {this.as,
      this.asn,
      required this.isProxy,
      this.timeZone,
      this.zipCode,
      this.longitude,
      this.latitude,
      this.cityName,
      this.countryCode,
      this.countryName,
      this.regionName,
      this.resIp,
      this.isp,
      this.domain,
      this.netSpeed,
      this.iddCode,
      this.areaCode,
      this.weatherStationCode,
      this.weatherStationName,
      this.mcc,
      this.mnc,
      this.mobileBrand,
      this.elevation,
      this.usageType,
      this.addressType,
      required this.continent,
      required this.country,
      required this.region,
      required this.city,
      required this.timeZoneInfo,
      required this.geotargeting});
}

class Continent {
  late String name;
  late String code;
  late List<String> hemisphere;
  late Translation translation; // Classe Translation
}

class Country {
  late String name;
  late String alpha3Code;
  late int numericCode;
  late String demonym;
  late String flag;
  late String capital;
  late int totalArea;
  late int population;
  late Currency currency; // Classe Currency
  late Language language; // Classe Language
  late String tld;
  late Translation translation; // Classe Translation
}

class Region {
  late String name;
  late String code;
  late Translation translation; // Classe Translation
}

class City {
  late String name;
  late Translation translation; // Classe Translation
}

class TimeZoneInfo {
  late String olson;
  late String currentTime;
  late int gmtOffset;
  late bool isDst;
  late String sunrise;
  late String sunset;
}

class Geotargeting {
  late String metro;
}

class Proxy {
  late int lastSeen;
  late String proxyType;
  late String threat;
  late String provider;
  late bool isVpn;
  late bool isTor;
  late bool isDataCenter;
  late bool isPublicProxy;
  late bool isWebProxy;
  late bool isWebCrawler;
  late bool isResidentialProxy;
  late bool isSpammer;
  late bool isScanner;
  late bool isBotnet;
}

class Currency {
  late String code;
  late String name;
  late String symbol;
}

class Language {
  late String code;
  late String name;
}

class Translation {
  late String lang;
  late String value;
}


lib/src/base_response_whois.dart
class HttpIkaWhoisService {
  ///
  late int? errorCode = 0;

  ///
  late String? errorMessage;

  /// Domain scanning
  late String? domain;

  /// Domain Id
  late String? domainId;

  ///Domain status
  late String? status;

  /// Create date
  late String? createDate;

  /// Updating date
  late String? updateDate;

  /// Expires date
  late String? expiresDate;

  /// Domain age
  late int? domainAge;

  /// Whois server
  late String? whoisServer;

  /// Registrar information
  late Registrar registrar;

  /// Registrant information
  late Registrant registrant;

  /// Admin information
  late Admin admin;

  ///Technical information
  late Tech technical;

  /// billing information
  late Billing billing;

  List<dynamic>? nameServer;

  HttpIkaWhoisService(
      {required this.admin,
      required this.technical,
      required this.billing,
      this.createDate,
      this.updateDate,
      this.domain,
      this.domainAge,
      this.domainId,
      this.expiresDate,
      required this.registrant,
      required this.registrar,
      this.status,
      this.whoisServer,
      this.nameServer});
}

class Registrar {
  String? ianaId;
  String? name;
  String? url;

  Registrar({
    this.ianaId,
    this.name,
    this.url,
  });
}

class Registrant {
  String? name;
  String? organization;
  String? streetAddress;
  String? city = "";
  String? region = "";
  String? zipCode = "";
  String? country = "";
  String? phone = "";
  String? fax = "";
  String? email = "";

  Registrant({
    this.name,
    this.organization,
    this.streetAddress,
    this.city,
    this.region,
    this.zipCode,
    this.country,
    this.phone,
    this.fax,
    this.email,
  });
}

class Admin {
  String? name;
  String? organization;
  String? streetAddress;
  String? city = "";
  String? region = "";
  String? zipCode = "";
  String? country = "";
  String? phone = "";
  String? fax = "";
  String? email = "";

  Admin({
    this.name,
    this.organization,
    this.streetAddress,
    this.city,
    this.region,
    this.zipCode,
    this.country,
    this.phone,
    this.fax,
    this.email,
  });
}

class Tech {
  String? name;
  String? organization;
  String? streetAddress;
  String? city = "";
  String? region = "";
  String? zipCode = "";
  String? country = "";
  String? phone = "";
  String? fax = "";
  String? email = "";

  Tech({
    this.name,
    this.organization,
    this.streetAddress,
    this.city,
    this.region,
    this.zipCode,
    this.country,
    this.phone,
    this.fax,
    this.email,
  });
}

class Billing {
  String? name;
  String? organization;
  String? streetAddress;
  String? city = "";
  String? region = "";
  String? zipCode = "";
  String? country = "";
  String? phone = "";
  String? fax = "";
  String? email = "";

  Billing({
    this.name,
    this.organization,
    this.streetAddress,
    this.city,
    this.region,
    this.zipCode,
    this.country,
    this.phone,
    this.fax,
    this.email,
  });
}


example/verifyip2location_example.dart
import 'package:verifyip2location/verifyip2location.dart' as verifyip2location;
import 'package:verifyip2location/verifyip2location.dart';

void main() async {
  // Your API key
  final apiKey = 'VOTRE_API_KEY';

  // ip2location class
  BaseHttpIkaResponse? ip2loc;

   // IP address you want to look up
  final ipAddress = '8.8.8.8';

  // Plan you want to use (e.g, 'free = 0', 'starter = 1', 'plus = 2', 'security = 3', etc.)
  int plan = 0;

  try {
    // Using the getIpgeo function to retrieve data
    ip2loc = await verifyip2location.getIpgeo(apiKey, ip: ipAddress, plan: plan, option: 'geoip');

    // Get information
    if (ip2loc.errorStatusCode == 10000 || ip2loc.errorStatusCode == 10001 ){
      print('Adresse IP: ${ip2loc.resIp}');
      print('Pays: ${ip2loc.countryName}');
      print('Ville: ${ip2loc.cityName}');
      print('Latitude: ${ip2loc.latitude}');
      print('Longitude: ${ip2loc.longitude}');
      print('As: ${ip2loc.as}');
      // And so on for other data you want to use.
    } else if (ip2loc.errorStatusCode == 10001){
      print('Error message: ' + ip2loc.errorReasonPhrase!);
    } else{
        print('Error message: ' + ip2loc.errorReasonPhrase!);
    }
  } catch (e) {
    print('An error occurred during the request : $e');
  }
}